### DISCUSSION TITLE
PopCount<T> and Generic Bit Counting for IBinaryInteger<T>

### DISCUSSION THREAD

**Discussion by BoCoKeith** [dotnet/runtime#86545] [Q&A] [answered]
I need to be able to get the number of bits _as an integer_ needed for a generic integer data type (IBinaryInteger<T>) inside a generic method.  I can't find a property or method that gives me this directly, am I missing something?

Assuming not, it seems like a great answer would be ...

    public int CountBits<T>() where T: IBinaryInteger<T> => (int)(T.PopCount(T.AllBitsSet));

... but the value returned from PopCount can't be cast to an integer.

My next choice is something like ...

    public int CountBits<T>() where T: IBinaryInteger<T> => T.AllBitsSet.GetByteCount() * 8;

... which is mostly fine, except that it wouldn't work for a integer type that isn't a multiple of 8 (say a hypothetical UInt7).

I'm left with counting the bits myself...

    public int CountBits<T>() where T: IBinaryInteger<T> {
        int count = 0;
        int value = T.AllBitsSet;
        while (value != T.Zero) {
            count++;
            value >>= 1;
        };
        return count;
    }

... which I don't mind (it works great), but I don't want to go down this path if there's a built-in way to do it.

As an aside, what is the logic behind having PopCount<T> return a T and not an int?  It seems like an odd choice with limited use cases (particular since other PopCount methods, such as those in BitConverter, do return ints).

**Comment by huoyaoyuan** [comment, marked as answer]
>  ... but the value returned from PopCount can't be cast to an integer.

Use `int.CreateXXX` to cast.
